core/src/main/java/org/infinispan/functional/impl/MetaParams.java

```diff
@@ -50,94 +51,183 @@
+   private <T extends MetaParam> T findNullable(Class<T> type) {
       for (MetaParam<?> meta : metas) {
-         if (meta.getClass().isAssignableFrom(type))
+         if (meta != null && meta.getClass().isAssignableFrom(type))
             return (T) meta;
       }
 
       return null;
    }
 
-   public void add(MetaParam.Writable meta) {
-      if (metas.length == 0)
+   public void add(MetaParam meta) {
+      assert meta != null;
+      if (metas.length == 0) {
          metas = new MetaParam[]{meta};
-      else {
-         boolean found = false;
+         length = 1;
+      } else {
+         int hole = -1;
          for (int i = 0; i < metas.length; i++) {
-            if (metas[i].getClass().isAssignableFrom(meta.getClass())) {
+            MetaParam<?> m = metas[i];
+            if (m == null) {
+               hole = i;
+            } else if (m.getClass().isAssignableFrom(meta.getClass())) {
                metas[i] = meta;
-               found = true;
+               assert checkLength();
+               return;
             }
          }
 
-         if (!found) {
+         ++length;
+         if (hole < 0) {
             MetaParam<?>[] newMetas = Arrays.copyOf(metas, metas.length + 1);
             newMetas[newMetas.length - 1] = meta;
             metas = newMetas;
+         } else {
+            metas[hole] = meta;
          }
+         assert checkLength();
+      }
+   }
+
+   private boolean checkLength() {
+      int l = 0;
+      for (MetaParam meta : metas) {
+         if (meta != null) ++l;
       }
+      return l == length;
    }
 
-   public void addMany(MetaParam.Writable... metaParams) {
-      if (metas.length == 0) metas = metaParams;
-      else {
+   public void addMany(MetaParam... metaParams) {
+      if (metas.length == 0) {
+         // Arrays in Java are convergent, therefore someone could pass MetaParam.Writable[] and
+         // we could try to store non-writable meta param in that array (or its copy) later.
+         if (metaParams.getClass().getComponentType() == MetaParam.class) {
+            metas = metaParams;
+         } else {
+            metas = Arrays.copyOf(metaParams, metaParams.length, MetaParam[].class);
+         }
+         length = (int) Stream.of(metas).filter(Objects::nonNull).count();
+      } else {
          List<MetaParam<?>> notFound = new ArrayList<>(metaParams.length);
-         for (MetaParam.Writable newMeta : metaParams) {
-            boolean found = false;
+         OUTER: for (MetaParam newMeta : metaParams) {
+            int hole = -1;
             for (int i = 0; i < metas.length; i++) {
-               if (metas[i].getClass().isAssignableFrom(newMeta.getClass())) {
+               MetaParam<?> m = metas[i];
+               if (m == null) {
+                  hole = i;
+               } else if (m.getClass().isAssignableFrom(newMeta.getClass())) {
                   metas[i] = newMeta;
-                  found = true;
+                  continue OUTER;
```

Comment:
I'd rather extract a separate method with the loop body and use `return`, if possible.